<html>
<head>
<!-- Latest compiled and minified CSS -->
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">

<!-- Optional theme -->
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap-theme.min.css" integrity="sha384-rHyoN1iRsVXV4nD0JutlnGaslCJuC7uwjduW9SVrLvRYooPp2bWYgmgJQIXwl/Sp" crossorigin="anonymous">

<!-- Latest compiled and minified JavaScript -->
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>
</head>

<style>
body { margin-left: 20px }
</style>
<body>
<h2><a href="../">CRISPOR</a> Changes</h2>
<p>Many thanks to the users that have contacted me, suggested features and tested them!</p>

<h4>Version 5.00, Jan 2022 - small fixes</h4>
<li>CasX target was on the wrong strand, Jason Fernandes, Scribetx Inc </li>
<li>On the sequence viewer, cleavage sites were shown at the wrong location when using Crispor on a new OSX with M1 CPU and Asian fonts, Taro Nakamura, Nat Inst Basic Biol Japan</li>

<h4>Version 4.99, April 2021 - more enzymes</h4>
<li>Added Inscripta MAD7 enzyme, C. Dustin Rubinstein, Univ of Wisconsin-Madison
<li>adding Thermocas9, Panji Banurea, Bogor Agricultural University, Indonesia
<li>adding SpRY, Ben Kleinstiver, Harvard
<li>adding mutation overview for the last 3bp of the guide, Tom Schreiber, IPB Halle, Germany
<li>Activated exon display, a lot of users have asked for this

<h4>Version 4.98, June 2020 - small improvements</h4>
<li>Better Geneious export, Brian Vogler, Bayer
<li>Better Snapgene export, Yarin Aknin, Hebrew University
<li>Fix saCas9 scaffold primers, Santiago Ch√°vez, Udelar, Uruguay
<li>Invisible work towords a future exon display, Max

<h4>Version 4.97, Dec 2019 - small bugfixes</h4>
<li>"Only exons" now works for tables with many off-targets, Thom Saunders, U Michigan
<li>Off-target primers fixed, thanks Francois Hafezi, Univ Luxemburg and Zuzanna Bukowy, Polish Acad of Sciences and 
<li>Allowing 21bp guides for Cpf1, thanks Suzy Markossian, ENS Lyon, France
<li>Fixing some repetitive guides not getting flagged as repetitive, big thanks to Josh Tycko

<h4>Version 4.96, Oct 2019 - changes to Cpf1 search</h4>
<li>Allowing TTTN as off-targets for Cpf1, removing Graf et al warnings for anything but NGG pams, thanks to Florian Hahn, Rothamsted Research, UK
<li>Fixing layout of guide in the sequence view which lead to PAMs being not under their sequence, thanks Alessandro Bellan, Uni Padova, Italy

<h4>Version 4.95, Sep 2019 - more output options, moving towards a better hg38 genome</h4>
<li>Excel output now has Graf columns and the option of adding all scores, thanks to Peter Stepper, OAW Vienna
<li>hg38 now defaults to the "analysis set" hopefully this will solve problems with _alts and the PAR region, thanks to Evgeni Frenkel, MIT
<li>The crispor.py command line version now accepts .bed input, thanks to Evgeni Frenkel, MIT


<h4>Version 4.93, August 8 2019 - better cDNA detection</h4>
<li>cDNA input is now better detected. Thanks to Imke De Grahl, Uni Hamburg

<h4>Version 4.92, June 23 2019 - Bugfix for Cpf1</h4>
<li>Fixed an old bug, which we had since CRISPOR 4. All PAMs that contain "V" were never found on the reverse strand. Thansk to Gemma Codner, MRC Harwell

<h4>Version 4.91, May 1st 2019 - Lindel predictions, Graf et al criteria</h4>
<ul>
<li>Finally fixed 'no genome' scoring option, thanks to Sanika Derebail, Harvard
<li>Added Lindel outcome prediction model, see <a href="https://www.biorxiv.org/content/10.1101/481069v1">preprint</a>, thanks to Wei Chen, UW
<li>Added Graf et al. most-inefficient-guide motifs, see <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6352712/">Graf et al Cell Reports 2019</a>, thanks to Frederic Flamant, ENS Lyon, France.
</ul>

<h4>Version 4.8, Mar 2019 - Queueing changes, Off-target primers</h4>
<ul>
<li>Adding primer parameters on off-target primer designer page. Thanks to Lian Liu, Tsinghua.
<li>BWASW on-target alignment is now part of the queued job. Thanks to 100+ people who complained about crashed jobs. :-)
<li>Added a CFD guide specificity score, thanks to Evgeni Frenkel, MIT
<li>Bae et al. are now cached and we have a new main column for outcome scores
<li>Fixed serious bug that made TTTV PAM almost unusable
</ul>

<h4>Version 4.7, Jan 2019 - PAR loci, CCtop </h4>
<ul>
<li>Added new enzyme NmeCas9. Thanks to Alireza Edraki, U Mass Med
<li>Treat human PAR regions as one. Thanks to Evgeni Frenkel, MIT
<li>Fixed the new CCtop efficiency score. Thanks to Florian LAFARGUETTE, Genoway
</ul>

<h4>Version 4.6, Nov 2018 - alternate loci </h4>
<ul>
<li>When off-targets are found on both an _alt sequence and a normal chromosome, the first 100% identical off-target on the _alt is ignored. Thanks to Jean-Paul.
</ul>

<h4>Version 4.5, Sep 2018 - saCas9 specificity scores</h4>
<ul>
<li>Improved the Saturating Mutagenesis Assistant for our CRISPOR ASHG Workshop in San Diego, Oct 17, 2018, with Neville Sanjana, Luca Pinello and Kendell Clement. It was great to see some of you there.
<li>Added saCas9 specificity from Tycko et al, many thanks to Josh Tycko for his source code. Nov 2018: fixed the score. 
<li>Added Enzyme ScanisCas9 PAM NNG: works with any base after the NNG but is high efficiency with T. Chatterjee et al, Science Advances, in press.  ScanisCas9 works with the same guide RNAs as SpyCas9.
<li>Added Enzyme iSpyMacCas9 PAM NAA: <a href="https://doi.org/10.1101/429654">Preprint</a>. iSpyMacCas9 works with the same guide RNAs as SpyCas9.
</ul>

<h4>Version 4.4, May 2018 - Cpf1/saCas9 scores</h4>
<ul>
<li>Showing distance to target from off-target, thanks to Joe Miano, Rochester.edu
<li>Allow custom prefix/suffix for the oligos in CRISPOR Batch, thanks to Yves Matthess, Uni Frankfurt
<li>More prominent warning for TTTT-containing guides directly in the table (thanks to Kis Adras, NAIK Hungary)
<li>Allow sort by offtarget count, better 'too many off-targets' warnings, thanks to James Eshleman
<li>Fixed table layout bug on Firefox, thanks to referee 2 of NAR paper
<li>Cpf1: added DeepCpf1 efficiency scores, thanks to Seonwoo Min at Soul National University for the help
<li>SaCas9 scoring: got Najm et al 2018 model, thanks to Mudra Hegde and John Doench, Broad Inst.
<li>A new CRISPOR manual, <a href="../manual/">here</a>
<li>Added xCas9, currently shows all the old spCas9 scores
<li>Changed the version of the Doench2016 score to the Azimuth implementation, to stay in sync with the Broad's website. This leads to very small differences. You can see the old scores by clicking "Show all"
</ul>

<h4>Version 4.3, Oct 2017 - CRISPOR Batch</h4>
<ul>
<li>New tool: CRISPOR Batch for lentiviral screens (suggested by Matt Canver)
<li>Primers are not designed in repeated regions anymore (thanks Yueh-Chiang Hu!)
<li>Fixing two bugs reported by Bernadette Neve: crash on many Ns in off-target, crash when no primers found in bulk primer design
</ul>

<h4>Version 4.2, June 2017 - Saturation mutagenesis</h4>
<ul>
<li>Saturation Mutagenesis Tool (Matt Canver, Luca Pinello)
<li>Genbank-like formats for various plasmid editors and Fasta export (Jean-Paul)
<li>Better command line support (Luca Pinello)
<li>PCR page: Fixed bad bug in T7 plasmid primer (many thanks to Macie Walker)
<li>PCR page: Changed lentiCrisprV2 enzyme Bbs1 -> BsmBI (thanks to Erin Simonds, UCSF)
<li>Fixed crash with unicode in batch name (thanks to Thom Saunders, U Mich)
<li>External websites can now pre-fill out the input form with values for the genome (parameter: "org") or a sequence (parameter: "seq"). Thanks to Patrick Lemaire, U Montpellier, and Frank Perez, INSERM Paris
<li>Fixed: Genbank download crash for unscored guides, adding better Geneious support (Alyce Chen, Harvard)
<li>Fixed: crash on primers for genomes without gene annotations (Hugues Renault, CNRS Strassburg)
</ul>

<h4>Version 4.0, Mar 2017 - Cpf1, Variants, sequence names, AddGene plasmids, UCSC tracks</h4>
<p>The following features were implemented. Many thanks to the people that have contacted me, suggested features and tested them!</p>
<ul>
<li>The whole site is around 10x faster now.  Mark Diekhans/UCSC fixed the twoBit tool I use, many thanks!
<li>Genome Variants can be overlaid onto the input sequence (Matthew Canver/Harvard, Jean-Paul Concordet)
<li>Support for the Cpf1 enzyme. Its PAM is TTTN. (Max)
<li>Input: Guides cannot overlap Ns in the sequence (Alberto Stolfi, NYU)
<li>Input: A name for the input sequence can now be shown on all pages (Jean-Paul)
<li>Main Page: Added many more restriction enzymes and highlighted on PCR product (Saulius Zuklys, Uni Basel)
<li>Main Page: Guides can be sorted by position (Virginie Virolle, Universite Nice)
<li>Main Page: Click OOF score to show Bae et al. microhomology (Tsuyoshi Momose, Villefranche)
<li>Main Page: Primers can be designed for off-targets and results exported to Crispresso (Max and Luca Pinello/Harvard)
<li>Main Page: Showing only the best two scores now, based on our paper. Others are optional. (Max) 
<li>Main Page: Sorting the off-targets by CFD score now (Max, Virginie Virolle)
<li>Main Page: Show only guides that start with G- or A- (Peter Rogowsky, ENS Lyon)
<li>Download: Adding sequence and version info to Excel file (Soren Warming/Max)
<li>PCR page: Fixed bad bug in T7 plasmid primer (many thanks to Macie Walker)
<li>PCR page: Dropdown to select the Addgene plasmid, changes the cloning sites (Jean-Paul Concordet)
<li>PCR Page: Tm and target length can be selected for the validation primers (Jean-Michel Heslan, Univ Nantes)
<li>PCR Page: Redesign of the input validation primers page (Max)
<li>PCR Page - more protocols: cloning in Ciona (Alberto), off-target sequencing (Matthew Canver, MIT), synthesis (Ann de Cian, MNHN Paris)
<li>UCSC Browser: Browser tracks with pre-computed guides for the major model organisms (Max)
<li>Local install: Off-target Excel export now includes the strand (Max). This changes the output format but was necessary.
<li>Local installation/offline-mode: all .js and .css files are now included in the distribution
<li>Local install: license is now free for academic use only. (Max)
</ul>

</body>
</html
